### PR TITLE
Update strcmp implementation, to improve performance

### PR DESCRIPTION
**Pull request by huxinx**
JerryScript-DCO-1.0-Signed-off-by: Xin Hu Xin.A.Hu@intel.com

binary size is 292k.
Run ./tools/run-perf-test.sh 10 times, 
OS, ubuntu 15.04, 32 bit
CPU, Intel(R) Celeron(R) CPU N2820 @ 2.13GHz, 2 core

                               Benchmark |         RSS<br>(+ is better) |        Perf<br>(+ is better) |
                               --------- |                          --- |                         ---- |
                              3d-cube.js |         132->   136 (-3.030) |     1.28667-> 1.288 (-0.103) |3d-morph.js | <FAILED3d-raytrace.js | <FAILED>
                  access-binary-trees.js |           88->    88 (0.000) |     0.676->0.677333 (-0.197) |
                      access-fannkuch.js |          48->    52 (-8.333) |     3.67333->3.62933 (1.198) |access-nbody.js | <FAILED>access-nsieve.js | <FAILED>
             bitops-3bit-bits-in-byte.js |           40->    40 (0.000) |        1.032->  1.02 (1.163) |
                  bitops-bits-in-byte.js |           44->    40 (9.091) |    1.45067->1.45467 (-0.276) |
                   bitops-bitwise-and.js |           40->    40 (0.000) |     1.46267->1.43333 (2.006) |bitops-nsieve-bits.js | <FAILED>
                controlflow-recursive.js |          244->   244 (0.000) |     0.633333->  0.62 (2.105) |
                           crypto-aes.js |          152->   152 (0.000) |        2.428->   2.4 (1.153) |
                           crypto-md5.js |          208->   208 (0.000) |      12.892->12.8387 (0.413) |
                          crypto-sha1.js |          148->   144 (2.703) |      5.63333-> 5.588 (0.805) |
                    date-format-tofte.js |          108->   108 (0.000) |     1.45867->1.44667 (0.823) |
                    date-format-xparb.js |          92->    96 (-4.348) |     0.762667-> 0.756 (0.874) |
                          math-cordic.js |           48->    48 (0.000) |     1.53733->1.50933 (1.821) |math-partial-sums.js | <FAILED>
                   math-spectral-norm.js |           44->    44 (0.000) |      0.848->0.834667 (1.572) |regexp-dna.js | <FAILED>
                        string-base64.js |         204->   208 (-1.961) |      87.1547->86.712 (0.508) |
                         string-fasta.js |           60->    60 (0.000) |     2.29467->2.27867 (0.697) |string-tagcloud.js | <FAILED>string-unpack-code.js | <FAILED>string-validate-input.js | <FAILED>
                         Geometric mean: |       RSS reduction: -0.308% |            Speed up: 0.9128% |

Update strcmp() implementation, simplify code, leverage from integer promotion, gets better performance.

The following is the assembly code on x86, 32bit platform,

 80534ba: 55                    push   %ebp
 80534bb: 89 e5                 mov    %esp,%ebp
 80534bd: 8a 08                 mov    (%eax),%cl
 80534bf: 3a 0a                 cmp    (%edx),%cl
 80534c1: 75 08                 jne    80534cb <strcmp+0x11>
 80534c3: 84 c9                 test   %cl,%cl
 80534c5: 74 0e                 je     80534d5 <strcmp+0x1b>
 80534c7: 40                    inc    %eax
 80534c8: 42                    inc    %edx
 80534c9: eb f2                 jmp    80534bd <strcmp+0x3>
 80534cb: 0f b6 12              movzbl (%edx),%edx
 80534ce: 0f b6 c1              movzbl %cl,%eax
 80534d1: 29 d0                 sub    %edx,%eax
 80534d3: eb 02                 jmp    80534d7 <strcmp+0x1d>
 80534d5: 31 c0                 xor    %eax,%eax
 80534d7: 5d                    pop    %ebp
 80534d8: c3                    ret

### CODE REVIEW

File: jerry-libc/jerry-libc.c [hunk 1]
```diff
@@ -158,41 +158,15 @@ CALL_PRAGMA (GCC diagnostic pop)
...
-    }
-  }
-
-  if (s2[i])
-  {
-    return -1;
-  }
 
-  return 0;
+  return (*(const unsigned char *)s1 - *(const unsigned char *)s2);
```
Comment (egavrin) [inline jerry-libc/jerry-libc.c]: `return (*s1 < *s2) ? -1 : (*s1 > *s2 ? 1 : 0);` ?

File: jerry-libc/jerry-libc.c [hunk 2]
```diff
@@ -158,41 +158,15 @@ CALL_PRAGMA (GCC diagnostic pop)
 int
 strcmp (const char *s1, const char *s2)
```
Comment (egavrin) [inline jerry-libc/jerry-libc.c]: BTW I'd prefer implementation that looks more like that:

```c
int strcmp(const char *str1, const char *str2)
{
    int s1;
    int s2;
    do {
        s1 = *str1++;
        s2 = *str2++;
        if (s1 == 0)
            break;
    } while (s1 == s2);
    return (s1 < s2) ? -1 : (s1 > s2);
}
```
Source: http://stackoverflow.com/a/29279634/230855

**Comment by tilmannOSG** [comment]
LGTM

I'm pretty sure the RSS increases are just noise.

File: jerry-libc/jerry-libc.c (hunk 1 above)
Comment (tilmannOSG) [inline jerry-libc/jerry-libc.c]: Not necessary, strcmp() does not need to return -1 or 1 specifically. The important part is the sign of the return value.

File: jerry-libc/jerry-libc.c (hunk 2 above)
Comment (tilmannOSG) [inline jerry-libc/jerry-libc.c]: This version is less efficient than Xin's implementation (it has an additional compare).

**Comment by huxinx** [comment]
Benchmark |         RSS<br>(+ is better) |        Perf<br>(+ is better) |
                               --------- |                          --- |                         ---- |
                              3d-cube.js |          136->   132 (2.941) |        1.292-> 1.288 (0.310) |3d-morph.js | <FAILED>3d-raytrace.js | <FAILED>
                  access-binary-trees.js |          84->    88 (-4.762) |         0.68-> 0.672 (1.176) |
                      access-fannkuch.js |           52->    48 (7.692) |       3.656-> 3.686 (-0.821) |access-nbody.js | <FAILED>access-nsieve.js | <FAILED>
             bitops-3bit-bits-in-byte.js |         40->    44 (-10.000) |        1.044-> 1.028 (1.533) |
                  bitops-bits-in-byte.js |           40->    40 (0.000) |         1.46-> 1.452 (0.548) |
                   bitops-bitwise-and.js |           44->    40 (9.091) |         1.47-> 1.434 (2.449) |bitops-nsieve-bits.js | <FAILED>
                controlflow-recursive.js |          244->   244 (0.000) |        0.624-> 0.608 (2.564) |
                           crypto-aes.js |          156->   156 (0.000) |       2.424-> 2.432 (-0.330) |
                           crypto-md5.js |          204->   204 (0.000) |       12.838->12.832 (0.047) |
                          crypto-sha1.js |         144->   148 (-2.778) |        5.598-> 5.596 (0.036) |
                    date-format-tofte.js |          108->   108 (0.000) |        1.502-> 1.448 (3.595) |
                    date-format-xparb.js |           96->    96 (0.000) |        0.752-> 0.752 (0.000) |
                          math-cordic.js |           48->    48 (0.000) |        1.508-> 1.508 (0.000) |math-partial-sums.js | <FAILED>
                   math-spectral-norm.js |           44->    44 (0.000) |        0.844-> 0.836 (0.948) |regexp-dna.js | <FAILED>
                        string-base64.js |         200->   204 (-2.000) |      86.446->86.534 (-0.102) |
                         string-fasta.js |           60->    60 (0.000) |         2.24-> 2.236 (0.179) |string-tagcloud.js | <FAILED>string-unpack-code.js | <FAILED>string-validate-input.js | <FAILED>
                         Geometric mean: |        RSS reduction: 0.101% |            Speed up: 0.7652% |


binary size is 292k.
Run ./tools/run-perf-test.sh 10 times, 
OS, ubuntu 15.04, 32 bit
CPU, Intel(R) Celeron(R) CPU N2820 @ 2.13GHz, 2 core

Seems the overflow version is fine on x86 too.

**Comment by tilmannOSG** [comment]
As mentioned before, I see little value in going with the version from Stack Overflow.

Code relying on strcmp() returning -1 or 1 is wrong and our implementation of strcmp() should expose those bugs - not hide them.

BTW the overall performance improvement you are seeing is unlikely to be related to the different implementation of strcmp(), it's more likely to be noise and/or performance differences due to a different code layout. At the moment JerryScript only uses strcmp() for command line parsing in main-linux.cpp and main-darwin.cpp. Thus the performance impact of a faster strcmp() implementation is rather limited :)

**Comment by ruben-ayrapetyan** [comment]
Looks good to me

**Comment by galpeter** [comment]
lgtm

**Comment by egavrin** [comment]
@ruben-ayrapetyan could you please merge it for @huxinx

**Comment by ruben-ayrapetyan** [comment]
@egavrin, yes, sure

**Comment by ruben-ayrapetyan** [comment]
Merged (ba95cb010ab7a5fee516e6a0555feec6ea67234e)

**Pull request closed without merging**